### PR TITLE
amp-kaltura-player | Add Kaltura player to the AMP project

### PR DESCRIPTION
**Pull request by itaykinnrot**

### CODE REVIEW

File: examples/kaltura.amp.html [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  <link href='https://fonts.googleapis.com/css?family=Georgia|Open+Sans|Roboto' rel='stylesheet' type='text/css'>
+  <script async custom-element="amp-kaltura" src="https://cdn.ampproject.org/v0//amp-kaltura-0.1.js"></script>
+  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
+  <script async src="https://cdn.ampproject.org/v0.js"></script>
+</head>
+<body>
+  <h2>Kaltura Player</h2>
+
+  <amp-kaltura
```
Comment (cramforce) [inline examples/kaltura.amp.html]: Rename to `amp-kaltura-player`

File: extensions/amp-kaltura/0.1/amp-kaltura.js [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+    /** @override */
+    layoutCallback() {
+      const width = this.element.getAttribute('width');
+      const height = this.element.getAttribute('height');
+      const partnerid = AMP.assert(
+            this.element.getAttribute('data-partner'),
+            'The data-account attribute is required for <amp-kaltura> %s',
+            this.element);
+      const uiconfid = (this.element.getAttribute('data-uiconf') ||
```
Comment (cramforce) [inline extensions/amp-kaltura/0.1/amp-kaltura.js]: nit: Here and elsewhere remove superfluous parentheses.

File: extensions/amp-kaltura/0.1/amp-kaltura.js [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+      this.element.getAttribute('data-uiconf-id') ||
+        'default');
+      const entryid = (this.element.getAttribute('data-entryid') || 'default');
+      const iframe = document.createElement('iframe');
+      let src = `https://cdnapisec.kaltura.com/p/${encodeURIComponent(partnerid)}/sp/${encodeURIComponent(partnerid)}00/embedIframeJs/uiconf_id/${encodeURIComponent(uiconfid)}/partner_id/${encodeURIComponent(partnerid)}?iframeembed=true&playerId=kaltura_player_amp&entry_id=${encodeURIComponent(entryid)}`;
+      const params = {};
+
```
Comment (cramforce) [inline extensions/amp-kaltura/0.1/amp-kaltura.js]: Nit: Remove vertical whitespace

File: extensions/amp-kaltura/0.1/amp-kaltura.js [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    /** @override */
+    isLayoutSupported(layout) {
+      return isLayoutSizeDefined(layout);
+    }
+
+    /** @override */
+    layoutCallback() {
+      const width = this.element.getAttribute('width');
```
Comment (cramforce) [inline extensions/amp-kaltura/0.1/amp-kaltura.js]: Do you have a way to draw a poster image?

**Comment by itaykinnrot** [comment]
Hi @cramforce - Thanks for the review - i've added fixes and re-merged master in.

**Comment by cramforce** [comment]
Could you respond to my question about poster images? Otherwise this looks good. Please squash commits.

**Comment by itaykinnrot** [comment]
@cramforce Squashed  - regarding the image - when the player is loading - we show a poster of the video before the user is pressing the play btn.

**Comment by cramforce** [comment]
@itaykinnrot Yep, my request would be to show that image before showing the iframe. See #2467 for an example.

**Comment by itaykinnrot** [comment]
@cramforce  added an image before the iframe - looks great! Thanks for the comment and the example.

File: extensions/amp-kaltura-player/0.1/amp-kaltura-player.js [hunk 5]
```diff
@@ -0,0 +1,125 @@
...
+        this.buildImagePlaceholder_();
+      }
+    }
+
+    /** @override */
+    layoutCallback() {
+      const width = this.element.getAttribute('width');
+      const height = this.element.getAttribute('height');
+      const partnerid = AMP.assert(
+        this.element.getAttribute('data-partner'),
```
Comment (cramforce) [inline extensions/amp-kaltura-player/0.1/amp-kaltura-player.js]: Line continuations like this should be indented 4 spaces. Sorry for not catching this in the linter.

File: extensions/amp-kaltura-player/0.1/amp-kaltura-player.js [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+      this.iframe_ = iframe;
+      return loadPromise(iframe);
+    }
+
+    /** @private */
+    buildImagePlaceholder_() {
+      const imgPlaceholder = new Image();
+
+      setStyles(imgPlaceholder, {
+            //Cover matches YouTube Player styling.
```
Comment (cramforce) [inline extensions/amp-kaltura-player/0.1/amp-kaltura-player.js]: Fix comment

File: extensions/amp-kaltura-player/0.1/amp-kaltura-player.js [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+import {isLayoutSizeDefined} from '../../../src/layout';
+import {loadPromise} from '../../../src/event-helper';
+import {addParamsToUrl} from '../../../src/url';
+import {dashToCamelCase} from '../../../src/string';
+import {setStyles} from '../../../src/style';
+
+class AmpKaltura extends AMP.BaseElement {
+
+    /** @override */
```
Comment (cramforce) [inline extensions/amp-kaltura-player/0.1/amp-kaltura-player.js]: Nit: indent -2

File: extensions/amp-kaltura-player/0.1/amp-kaltura-player.js [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+        'visibility': 'hidden',
+      });
+      const width = this.element.getAttribute('width');
+      const height = this.element.getAttribute('height');
+      const partnerid = AMP.assert(
+        this.element.getAttribute('data-partner'),
+        'The data-account attribute is required for <amp-kaltura-player> %s',
+        this.element);
+      const entryid = this.element.getAttribute('data-entryid') || 'default';
+      let src = `https://cdnapisec.kaltura.com/p/${encodeURIComponent(partnerid)}/thumbnail/entry_id/${encodeURIComponent(entryid)}`;
```
Comment (cramforce) [inline extensions/amp-kaltura-player/0.1/amp-kaltura-player.js]: Just making sure: Is this the exact same URL the player will load for its poster?

File: extensions/amp-kaltura-player/0.1/test/test-amp-kaltura-player.js [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+    }, true).then(bc => {
+      const iframe = bc.querySelector('iframe');
+      expect(iframe).to.not.be.null;
+      expect(iframe.className).to.match(/-amp-fill-content/);
+    });
+  });
+
+  it('requires data-account', () => {
+    return getKaltura({}).should.eventually.be.rejectedWith(
+            /The data-account attribute is required for/);
```
Comment (cramforce) [inline extensions/amp-kaltura-player/0.1/test/test-amp-kaltura-player.js]: nit: should be 4 spaces indented from the `return`

File: extensions/amp-kaltura-player/0.1/test/test-amp-kaltura-player.js [hunk 10]
```diff
@@ -0,0 +1,89 @@
...
+  });
+
+    // TODO(erwinm) unskip this when we figure out why it fails on travis
+  it.skip('should pass data-param-* attributes to the iframe src', () => {
```
Comment (cramforce) [inline extensions/amp-kaltura-player/0.1/test/test-amp-kaltura-player.js]: Not skip? :)

**Comment by cramforce** [comment]
Looks great. A few comments.

**Comment by cramforce** [comment]
One more thing: Please add a `documentInactiveCallback` to your element. This should stop the video when the player gets swiped away. Most players do this with a `postMessage`. If you do not support this, you need to remove the iframe.

**Comment by itaykinnrot** [comment]
@cramforce fixes and added pause on inactive (our player support it) - Thanks

File: extensions/amp-kaltura-player/0.1/amp-kaltura-player.js [hunk 11]
```diff
@@ -0,0 +1,133 @@
+/**
+ * Copyright 2015 The AMP HTML Authors. All Rights Reserved.
```
Comment (erwinmombay) [inline extensions/amp-kaltura-player/0.1/amp-kaltura-player.js]: 2016 here and every other new file

**Comment by itaykinnrot** [comment]
update to 2016

**Comment by dvoytenko** [comment]
@itaykinnrot Could you please look in the https://travis-ci.org/ampproject/amphtml/builds/115828531 and fix the lint errors? You can simply run `gulp lint` to see these messages on your local. Otherwise this looks ready to be merged - just need to get Travis green.

**Comment by itaykinnrot** [comment]
@dvoytenko fixed

**Comment by dvoytenko** [comment]
Thanks! LGTM. The errors are unrelated and already fixed, so I'm merging this PR.

**Pull request merged**